src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,345 @@
+        {
+            int maxAttempts = 5;
+            int attempts = 0;
+            int leasesToAquire = expectedTotalLeaseCount - this.myQueues.Count;
+            this.logger.Info($"AcquireLeasesToMeetExpection : Try to acquire {leasesToAquire} queues");
+            while (attempts ++ <= maxAttempts && leasesToAquire > 0)
+            {
+                //select new queues to acquire
+                var expectedQueues = this.queueSelector.NextSelection(leasesToAquire, this.myQueues.Select(queue=>queue.QueueId)).ToList();
+                var leaseRequests = expectedQueues.Select(queue => new LeaseRequest() {
+                    ResourceKey = queue.ToString(),
+                    Duration = this.leaseLength
+                });
+                var results = await this.leaseProvider.Acquire(ResourceCategory.Streaming, leaseRequests.ToArray());
+                //add successfully acquired queue to myQueues list
+                for (int i = 0; i < results.Count(); i++)
+                {
+                    if (results[i].StatusCode == ResponseCode.OK)
+                    {
+                        this.myQueues.Add(new AcquiredQueue(expectedQueues[i], results[i].AcquiredLease));
+                    }
+                }
+                //if reached expectedTotalLeaseCount
+                if (this.myQueues.Count >= expectedTotalLeaseCount)
+                {
+                    break;
+                }
+            }
+
+            this.logger.Info($"AcquireLeasesToMeetExpection: finished. Now own {this.myQueues.Count} queues. Used attemps : {attempts}, Current minimumReponsibility : {this.minimumResponsibilty}, current maximumResponsibility : {this.maximumRespobsibility}");
+        }
+        
+        private async Task RenewLeases()
+        {
+            var results = await this.leaseProvider.Renew(ResourceCategory.Streaming, this.myQueues.Select(queue => queue.AcquiredLease).ToArray());
+            var updatedQueues = new List<AcquiredQueue>();
+            //update myQueues list with successfully renewed leases
+            for (int i = 0; i < results.Count(); i++)
+            {
+                var result = results[i];
+                if (result.StatusCode == ResponseCode.OK)
+                {
+                    updatedQueues.Add(new AcquiredQueue(this.myQueues[i].QueueId, result.AcquiredLease));
+                }
+            }
+            this.myQueues.Clear();
+            this.myQueues = updatedQueues;
+            this.logger.Info($"RenewLeases: finished, currently own queues : {this.myQueues.Count}, current minimumResponsibilty : {this.minimumResponsibilty}, current maximunResponsibilty : {this.maximumRespobsibility}");
+        }
+
+        private void CalculateResponsibility()
+        {
+            int activeBuckets = 0;
+            if (isStarting)
+            {
+                activeBuckets = this.deploymentConfig.GetAllSiloNames().Count;
```

Comment:
Are we confident the list of silo names will properly reflect deployment in autoscale environment? I suspect buckets should be the number of active silos in the cluster, regardless of deployment configuration.